### ISSUE TITLE
Breakage in RWO since 1.6

### ISSUE THREAD

**Issue by NathanReb** [realworldocaml/mdx#265]
I ran some tests with our master branch against RWO and it seems that we broke some of the toplevel blocks/envs behaviour from the look of it.

The test branch can be found here: https://github.com/NathanReb/book/tree/test-new-mdx. Some of the tests in the Classes and FFI are broken with master whereas they worked fine with 1.6.

We need to look at this before being able to release a new version of MDX!

**Comment by NathanReb** [comment]
`dune runtest` will work, you can also test individual chapters with `dune runtest book/<chapter>`!

**Comment by gpetiot** [comment]
It seems the semantics of preludes has changed:
- then: `--prelude foo.ml` would apply the prelude for each environment
- now: `--prelude foo.ml` applies the prelude only for the default environment, we need `--prelude bar:foo.ml` to apply the prelude to environment bar

The new preludes stanza does not allow the bar:foo.ml syntax though

**Comment by gpetiot** [comment]
Applying the following diff helps if we don't want to change the semantics again:
```diff
diff --git a/book/classes/dune b/book/classes/dune
index 5cf1868b..7b5520a2 100644
--- a/book/classes/dune
+++ b/book/classes/dune
@@ -4,6 +4,10 @@
   core
   mdx
   ppx_jane)
- (preludes prelude.ml))
+ (preludes
+   (env binary prelude.ml)
+   (env istack prelude.ml)
+   (env iter prelude.ml)
+   (env stack prelude.ml)))
 
 (data_only_dirs examples)
diff --git a/book/command-line-parsing/dune b/book/command-line-parsing/dune
index 5cf1868b..160b1ce6 100644
--- a/book/command-line-parsing/dune
+++ b/book/command-line-parsing/dune
@@ -4,6 +4,7 @@
   core
   mdx
   ppx_jane)
- (preludes prelude.ml))
+ (preludes
+   (env main prelude.ml)))
 
 (data_only_dirs examples)
diff --git a/book/data-serialization/dune b/book/data-serialization/dune
index 5cf1868b..3f6aeb92 100644
--- a/book/data-serialization/dune
+++ b/book/data-serialization/dune
@@ -4,6 +4,16 @@
   core
   mdx
   ppx_jane)
- (preludes prelude.ml))
+ (preludes
+   (env sexp_printer prelude.ml)
+   (env to_from_sexp prelude.ml)
+   (env example_load prelude.ml)
+   (env sexp_opaque prelude.ml)
+   (env manually_making_sexp prelude.ml)
+   (env auto_making_sexp prelude.ml)
+   (env inline_sexp prelude.ml)
+   (env sexp_list prelude.ml)
+   (env sexp_option prelude.ml)
+   (env sexp_default prelude.ml)))
 
 (data_only_dirs examples)
diff --git a/book/first-class-modules/dune b/book/first-class-modules/dune
index 5cf1868b..398961e2 100644
--- a/book/first-class-modules/dune
+++ b/book/first-class-modules/dune
@@ -4,6 +4,7 @@
   core
   mdx
   ppx_jane)
- (preludes prelude.ml))
+ (preludes
+   (env query_handler prelude.ml)))
 
 (data_only_dirs examples)
diff --git a/book/foreign-function-interface/dune b/book/foreign-function-interface/dune
index 95edb682..0f09468f 100644
--- a/book/foreign-function-interface/dune
+++ b/book/foreign-function-interface/dune
@@ -7,6 +7,7 @@
   mdx
   ocaml-print-intf
   ppx_jane)
- (preludes prelude.ml))
+ (preludes
+   (env posix prelude.ml)))
 
 (data_only_dirs examples)
diff --git a/book/json/dune b/book/json/dune
index 5e12ae4e..65abe0f1 100644
--- a/book/json/dune
+++ b/book/json/dune
@@ -5,6 +5,7 @@
   mdx
   ppx_jane
   yojson)
- (preludes prelude.ml))
+ (preludes
+   (env parse_book prelude.ml)))
 
 (data_only_dirs examples)
diff --git a/book/objects/dune b/book/objects/dune
index 8bd00edc..6f35cab2 100644
--- a/book/objects/dune
+++ b/book/objects/dune
@@ -5,7 +5,7 @@
   mdx
   ppx_jane)
  (preludes
-  prelude.ml
-  subtyping.ml))
+   (env row_polymorphism prelude.ml)
+   (env subtyping subtyping.ml)))
 
 (data_only_dirs examples)
diff --git a/book/ppx/dune b/book/ppx/dune
index f8905ba4..e47e148c 100644
--- a/book/ppx/dune
+++ b/book/ppx/dune
@@ -4,4 +4,5 @@
   core
   mdx
   ppx_jane)
- (preludes prelude.ml))
+ (preludes
+   (env main prelude.ml)))
diff --git a/book/variants/dune b/book/variants/dune
index b2f60c0d..9edf95fe 100644
--- a/book/variants/dune
+++ b/book/variants/dune
@@ -4,6 +4,8 @@
   core
   mdx
   ppx_jane)
- (preludes prelude.ml old_termcolor.ml))
+ (preludes
+   (env main prelude.ml)
+   (env old_termcolor old_termcolor.ml)))
 
 (data_only_dirs examples)
```

**Comment by gpetiot** [comment]
Closing this one since the prelude issue has been fixed in #271, see #272 for other regressions
